Fix normalization of np.str_ and np.bytes_ types

yashwantbezawada opened the pull request:
#### Reference Issues/PRs

Fixes #2800

#### What does this implement or fix?

When you try to write a DataFrame with `np.str_` values, you get this error:

```
ArcticDbNotYetImplemented: Failed to normalize column 'col' with dtype 'object'. 
Found first non-null value of type '<class 'numpy.str_'>', but only strings, unicode, 
and Timestamps are supported.
```

The issue is in `_accept_array_string()` - it checks `type(v) in (str, bytes)` which does an exact type match. But `np.str_` is its own class that inherits from `str`, so the check fails even though it is basically a string.

Switched to `isinstance(v, (str, bytes))` which handles subclasses properly. Same thing for `coerce_string_column_to_fixed_length_array()` - changed `to_type == str` to `issubclass(to_type, str)`.

#### Any other comments?

Quick sanity check that this makes sense:
```python
>>> isinstance(np.str_("hello"), str)
True
>>> isinstance(np.bytes_(b"hello"), bytes)
True
```

The C++ side already does the right thing - `PyUnicode_Check` and `PyBytes_Check` both check the type hierarchy, so `np.str_` values work fine once they get past the Python normalization.

Added some tests to cover this.

#### Checklist

<details>
  <summary>
   Checklist for code changes...
  </summary>
 
- [x] Have you updated the relevant docstrings, documentation and copyright notice?
- [x] Is this contribution tested against [all ArcticDB's features](../docs/mkdocs/docs/technical/contributing.md)?
- [x] Do all exceptions introduced raise appropriate [error messages](https://docs.arcticdb.io/error_messages/)?
- [x] Are API changes highlighted in the PR description?
- [x] Is the PR labelled as enhancement or bug so it appears in autogenerated release notes?
</details>

yashwantbezawada committed:
Fix normalization of np.str_ and np.bytes_ types

The normalization code was using exact type checks (type(v) in (str, bytes))
which failed for numpy string scalar types since np.str_ and np.bytes_ are
distinct classes that inherit from str and bytes respectively.

Changed to use isinstance() checks which properly handle the type hierarchy,
allowing numpy string types to be normalized correctly.

Also updated coerce_string_column_to_fixed_length_array to use issubclass()
for consistent handling when dynamic_strings=False.

Added regression tests for the fix.

Fixes #2800

Signed-Off By: Yashwant Bezawada <yashwant_b@me.com>. By including this sign-off line I agree to the terms of the Contributor License Agreement.

IvoDD commented on python/arcticdb/version_store/_normalization.py:
```diff
@@ -157,7 +157,8 @@ def from_npd_df(df):
     # and does not transform string into bytes
-    return type(v) in (str, bytes)
+    # Use isinstance to also support numpy string types (np.str_, np.bytes_)
+    return isinstance(v, (str, bytes))
```
This will accept any type that inherits `str`, which I think is also a good change.

What do you think about putting that behavior in another test like:
```
    class CustomStrType(str, Enum):
        A = "A"
        B = "B"

    # Test custom string types in DataFrame
    df = pd.DataFrame({"col": [CustomStrType.A, CustomStrType.B]})
    lib.write(sym, df)
    result = lib.read(sym).data
    # Custom string values should be read back as regular strings
    assert result["col"][0] == "A"
    assert result["col"][1] == "B"
```

IvoDD reviewed (commented):
Thank you for contributing! Looks good! Just one suggestion for an extra test

yashwantbezawada committed:
Use strict type equality instead of isinstance for string type checks

Per reviewer feedback referencing issue #704, using isinstance() to
accept all str/bytes subclasses has been problematic before. Changed
to use strict type equality with explicit numpy types:
- type(v) in (str, bytes, np.str_, np.bytes_)
- to_type in (str, np.str_)

This supports numpy string types while avoiding issues with arbitrary
string subclasses.
